### ISSUE TITLE
Cycling through tabs moves in the wrong order if tabs are reordered

### ISSUE THREAD

**Issue by lexi-lambda** [racket/drracket#52]
The changes in racket/gui@9ef883a79f8d1a63fa2514183184e1741e83a181 allow tabs to be reordered on OS X, which is nice! However, it breaks the behavior of the “Next Tab” and “Previous Tab” menu item/hotkeys, which cycle through tabs in the *order they were created* instead of the current order after moving them around.

I’m not sure if this is an issue with `racket/gui`, DrRacket, or even `framework`, but I’m reporting it here since it’s how I ran into the bug, and I don’t see any obvious interface to cycle through tabs programmatically to test it myself.

**Comment by soegaard** [comment]
Most likely related:  cmd+9 no longer jumps to the last tab.

**Comment by mflatt** [comment]
It's not a feature; it's a bug. :)

Allowing tab reordering is not intended. Probably I need to set a flag explicitly to disable reordering for MMTabBarView and the default was the opposite for PSMTabBarControl. So, I think the short-term fix is to disable reordering for the new widget in `racket/gui`.

I would like to add support for tab reordering and close boxes, but it will take a while to sort through the capabilities and interfaces of the three to five native APIs, and I haven't made time for that so far.

**Comment by rfindler** [comment]
Ah, I see Matthew beat me to it, but here's a program that illustrates what
DrRacket is doing. FWIW.

```
#lang racket/gui
(define f (new frame% [label ""] [width 500] [height 100]))
(define panels '("first" "second" "third" "fourth"))
(define tp (new tab-panel%
                (font small-control-font)
                (parent f)
                (stretchable-height #f)
                (style '(no-border))
                (choices panels)))
(for ([x (in-list panels)]
      [i (in-naturals)])
  (new button%
       [parent f]
       [label (~a i)]
       [callback (λ args (send tp set-selection i))]))
(send f show #t)
```

**Comment by rfindler** [comment]
Ah, I see Matthew beat me to it, but here's a program that illustrates what
DrRacket is doing. FWIW.

```
#lang racket/gui
(define f (new frame% [label ""] [width 500] [height 100]))
(define panels '("first" "second" "third" "fourth"))
(define tp (new tab-panel%
                (font small-control-font)
                (parent f)
                (stretchable-height #f)
                (style '(no-border))
                (choices panels)))
(for ([x (in-list panels)]
      [i (in-naturals)])
  (new button%
       [parent f]
       [label (~a i)]
       [callback (λ args (send tp set-selection i))]))
(send f show #t)
```

**Comment by mflatt** [comment]
@soegaard Do you mean that Cmd-9 doesn't work after at least 9 tabs are created, even without reordering? Or that it doesn't select the last tab after reordering? Or something else?

**Comment by soegaard** [comment]
2016-04-18 4:58 GMT+02:00 Matthew Flatt <notifications@github.com>:

> @soegaard <https://github.com/soegaard> Do you mean that Cmd-9 doesn't
> work after at least 9 tabs are created, even without reordering? Or that it
> doesn't select the last tab after reordering? Or something else?
>
@mflatt  I think of cmd-9 as "go to last tab". When I tried it, I had 5
tabs open, and cmd-9 did nothing.
I usually have lots of tabs open, so I am unsure whether it actually worked
like that before the
new tabs.

/Jens Axel

**Comment by mflatt** [comment]
@soegaard Based on my experiments with v6.4, I don't think Cmd-9 worked that way before (i.e., it would not select the last tab if only 5 tabs were open). As long as Cmd-9 still works for you when you have lots of tabs open, I think the new tabs behave the same as before.

**Comment by soegaard** [comment]
2016-04-18 14:38 GMT+02:00 Matthew Flatt <notifications@github.com>:

> @soegaard <https://github.com/soegaard> Based on my experiments with
> v6.4, I don't think Cmd-9 worked that way before (i.e., it would not select
> the last tab if only 5 tabs were open). As long as Cmd-9 still works for
> you when you have lots of tabs open, I think the new tabs behave the same
> as before.
>
A quick experiment in Google Chrome with two tabs reveals that cmd-9 jumps
to the last tab.
The cmd-9 is the last tab idea is probably from Chrome then.

/Jens Axel

**Comment by rfindler** [comment]
Thanks for pointing this out. I've fixed DrRacket: https://github.com/racket/drracket/commit/d0f44e1b514ab46d62c0b76ea2b1945e0fde983c

**Comment by soegaard** [comment]
Thanks!

2016-04-18 15:00 GMT+02:00 Robby Findler <notifications@github.com>:

> Thanks for pointing this out. I've fixed DrRacket: d0f44e1
> <https://github.com/racket/drracket/commit/d0f44e1b514ab46d62c0b76ea2b1945e0fde983c>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/racket/drracket/issues/52#issuecomment-211370705>
>



-- 
-- 
Jens Axel Søgaard
